numba/targets/tupleobj.py

```diff
@@ -380,3 +382,98 @@ def tuple_index_impl(tup, value):
+#Intrinsics
+
+@intrinsic
+def make_tuple(typingctx, size, iterable):
+    """Creates a tuple of predetermined size from an iterable.
+
+    Parameters
+    ----------
+    size : IntegerLiteral
+        size of the tuple to be created
+    iterable : Iterable
+        iterable from which to draw the data with at least `size` elements
+
+    Returns
+    -------
+    UniTuple(typ, size)
+        tuple with contents of iterable where `typ` is the type of the contents
+        of the iterable
+
+    Raises
+    ------
+    RequireLiteralValue
+        raised during typing if 'size' parameter is not a constant integer
+    TypeError
+        raised during typing if iterable is of an invalid type
+    ValueError
+        raised at runtime if `len(iterable) < size`
+    """
+    if not isinstance(size, types.IntegerLiteral):
+        raise RequireLiteralValue(
+            "make_tuple: argument 'size' must be a constant integer")
+    if not isinstance(iterable, types.IterableType):
+        raise TypeError(
+            "make_tuple: argument 'iterable' must be an iterable")
+    if isinstance(iterable, types.Array):
+        if not iterable.ndim == 1:
+            raise TypeError(
+                "make_tuple: arrays provided as 'iterable' must be 1d")
+    if hasattr(iterable, 'dtype'):
+        dtype = iterable.dtype
+    elif hasattr(iterable, 'yield_type'):
+        dtype = iterable.yield_type
+    else:
+        raise TypeError(
+            f"make_tuple: cannot determine dtype of argument 'iterable'"
+            " of type '{iterable}'")
+
+    tuple_size = size.literal_value
+    tuple_type = types.UniTuple(dtype, tuple_size)
+    sig = tuple_type(size, iterable)
+
+    # Separate implementations are required to ensure that for iterators
+    # multiple calls continue the iteration while multiple calls
+    # with array, lists, etc must restart at the beginning
+    too_short_msg = "make_tuple: argument 'iterable' is shorter than specified"
+    if isinstance(iterable, (types.IteratorType, types.SimpleIterableType)):
+        def make_tuple_impl(iterable, empty_tuple):
+            out = empty_tuple
+            i = 0
+            for i, val in enumerate(iterable):
+                out = tuple_setitem(out, i, val)
+                if i+1 == tuple_size:
+                    break
+            if i+1 != tuple_size:
+                raise ValueError(too_short_msg)
+            return out
+    else:
+        def make_tuple_impl(iterable, empty_tuple):
+            out = empty_tuple
+            i = 0
+            if len(iterable) < tuple_size:
+                raise ValueError(too_short_msg)
+            for i in range(tuple_size):
+                out = tuple_setitem(out, i, iterable[i])
+            return out
+
+    def codegen(context, builder, signature, args):
+        [_, iterable_ty] = sig.args
+        [_, iterable] = args
+
+        # Allocate an empty tuple
+        empty_tuple = context.get_constant_undef(tuple_type)
```

Comment:
I think this has to be a `.get_constant_null` to zero-fill in case the tuple contains refcounted types.

Reply:
@sklam I am sorry but I have no idea. I will need some definite guidance here. You want me to make the change just to be safe or is there a way I can test its necessity?

Reply:
@sklam I looked into this myself and to be honest I cannot follow your reasoning. If I understand correctly `get_constant_undef()` does something akin to pure allocation while `get_constant_null()` does additinally zero fill, right?

Since all slots in the tuple will be filled any zero-filling is redundant. Anything not filling the tuple completely would raise an error.

The null-initialization seems to get optimized away. The resulting llvm is identical whether I use `get_constant_undef()` or `get_constant_null()`.

That said, you were specifically mentioning refcounted types as the possible problem. I do not know which they are and thus could not test this case specifically.

Do you want me to just use `get_constant_null()` to be safe and be done with it? After all it does not seem to hurt in the case of non-refcounted types.

---
This is the test-code I used:

```python
from numba import njit
from numba.cpython.tupleobj import make_tuple

@njit
def test_tuple_creation():
    return make_tuple(5, range(5))

test_tuple_creation()
print(test_tuple_creation.inspect_llvm(test_tuple_creation.signatures[0]))
```